cluster-autoscaler/core/loop_state.go

```diff
@@ -0,0 +1,448 @@
+	CanContinue ExecutionResult = iota
+	// ShouldStop means that the processing should be stopped and error (or nil) returned
+	ShouldStop
+)
+
+// LoopStateHandler hides complexity of Autoscaler.RunOnce implementation,
+// splitting it into a number of phases.
+// Each phase method returns ExecutionResult to show if processing can continue further
+// and a result (error or nil) to return from Autoscaler.RunOnce if the processing should stop.
+// NOTE: Do not use `err != nil` idiom to decide if the processing can continue.
+// LoopStateHandler instance is a very short-lived one, persisting state of a single RunOnce call.
+type LoopStateHandler interface {
+	// RefreshState obtains new Node-level data from the cloud provider
+	// and updates internal ClusterStateRegistry, until metrics.UpdateState.
+	// See LoopStateHandler doc for description of return values.
+	RefreshState() (ExecutionResult, errors.AutoscalerError)
+	// CheckPreconditions verifies Node-level preconditions for continuing
+	// to processing Pod-level data.
+	// See LoopStateHandler doc for description of return values.
+	CheckPreconditions() (ExecutionResult, errors.AutoscalerError)
+	// PreparePodData obtains Pod-level data for further processing.
+	// See LoopStateHandler doc for description of return values.
+	PreparePodData() (ExecutionResult, errors.AutoscalerError)
+	// ScaleUp executes scale-up, if it's required.
+	// See LoopStateHandler doc for description of return values.
+	ScaleUp() (ExecutionResult, errors.AutoscalerError)
+	// ScaleDown executes scale-down logic, both dry-run & actual scale-down, if it's required.
+	// See LoopStateHandler doc for description of return values.
+	ScaleDown() (ExecutionResult, errors.AutoscalerError)
+}
+
+// loopState is the default implementation of LoopStateHandler
+type loopState struct {
+	currentTime        time.Time
+	startTime          time.Time
+	scaleDownForbidden bool
+
+	autoscaler *StaticAutoscaler
+	context    *context.AutoscalingContext
+
+	allNodes                                       []*apiv1.Node
+	readyNodes                                     []*apiv1.Node
+	allScheduled                                   []*apiv1.Pod
+	unschedulablePodsToHelp                        []*apiv1.Pod
+	unschedulableWaitingForLowerPriorityPreemption []*apiv1.Pod
+}
+
+// NewLoopStateHandler creates new instance of the default LoopStateHandler implementation
+func NewLoopStateHandler(autoscaler *StaticAutoscaler, currentTime time.Time) *loopState {
+	return &loopState{
+		currentTime: currentTime,
+		startTime:   time.Now(),
+		autoscaler:  autoscaler,
+		context:     autoscaler.AutoscalingContext,
+	}
+}
+
+// See LoopStateHandler interface
+func (ls *loopState) RefreshState() (ExecutionResult, errors.AutoscalerError) {
+	if result, err := ls.refreshCloudProvider(); result != CanContinue {
+		return result, err
+	}
+	if result, err := ls.obtainNodeLists(); result != CanContinue {
```

Comment:
This also doesn't look good - why should obtainNodeList decide whether to continue with the main loop? The logic for handling empty cluster could easily be moved outside of it.

Reply:
Well, not so easily since the obtainNodeList has two exits on errors and two exits on business conditions, intertwined.
As discussed with @MaciekPytel we'd have to make sure that gpu.FilterOutNodesWithUnreadyGpus is safe to be called on empty allNodes list (it looks like it is but it's not tested for this) before reordering the conditions or just sticking with the final one.

The `result != CanContinue` construct is for both error handling and business decisions (here and in general). The method could return separate statuses for errors and business decisions but actually that won't affect this particular line of code.

Of course we could argue that obtainXYZ method should not have side effects and I'd happily agree there. But currently the logic is following the pattern seen prior to the refactor, I tried to avoid introducing functional changes, even if they are tiny (there are few, though, where they seemed to be bug-fixes).

I'm open for further discussion there - the goal of the refactor was not to end up with ideal code but just to move slightly further away from the monstrous RunOnce we've had.

In this particular case I can only see changing the method name or splitting it further, the latter will remove most of the gains we have. If you see a better way forward I'm happy to follow as I don't expect you want only a name change.